stratum/lib/p4runtime/sdn_controller_manager.h

```diff
@@ -0,0 +1,170 @@
+// Copyright 2020 Google LLC
+// Copyright 2021-present Open Networking Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef STRATUM_LIB_P4RUNTIME_SDN_CONTROLLER_MANAGER_H_
+#define STRATUM_LIB_P4RUNTIME_SDN_CONTROLLER_MANAGER_H_
+
+#include <string>
+#include <vector>
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/container/flat_hash_set.h"
+#include "absl/numeric/int128.h"
+#include "absl/status/status.h"
+#include "p4/v1/p4runtime.grpc.pb.h"
+#include "p4/v1/p4runtime.pb.h"
+
+namespace stratum {
+namespace p4runtime {
+
+// Named role for a SDN controller.
+constexpr char kP4RuntimeRoleSdnController[] = "sdn_controller";
+
+// A connection between a controller and p4rt server.
+class SdnConnection {
+ public:
+  SdnConnection(
+      grpc::ServerContext* context,
+      grpc::ServerReaderWriterInterface<p4::v1::StreamMessageResponse,
+                                        p4::v1::StreamMessageRequest>* stream)
+      : initialized_(false), grpc_context_(context), grpc_stream_(stream) {}
+
+  void Initialize() { initialized_ = true; }
+  bool IsInitialized() const { return initialized_; }
+
+  void SetElectionId(const absl::optional<absl::uint128>& id);
+  absl::optional<absl::uint128> GetElectionId() const;
+
+  void SetRoleName(const absl::optional<std::string>& name);
+  absl::optional<std::string> GetRoleName() const;
+
+  // A unique name string for the controller.
+  std::string GetName() const;
+
+  // Sends back StreamMessageResponse to this controller.
+  void SendStreamMessageResponse(const p4::v1::StreamMessageResponse& response);
+
+ private:
+  // The SDN connection should be initialized through arbitration before it can
+  // be used.
+  bool initialized_;
+
+  // SDN connections are made to the P4RT gRPC service based on role types. The
+  // specified role limits the table a connection can write to, and read from.
+  // If no role is specified then the connection is assumed to be root, and has
+  // access to all tables.
+  absl::optional<std::string> role_name_;
+
+  // Multiple connections can be established per role, but only one connection
+  // (i.e. the primary connection) is allowed to modify state. The primary
+  // connection is determined based on the election ID.
+  absl::optional<absl::uint128> election_id_;
+
+  // While the gRPC connection is open we keep access to the context & the
+  // read/write stream for communication.
+  grpc::ServerContext* grpc_context_;  // not owned.
+  grpc::ServerReaderWriterInterface<p4::v1::StreamMessageResponse,
+                                    p4::v1::StreamMessageRequest>*
+      grpc_stream_;  // not owned.
+};
+
+class SdnControllerManager {
+ public:
+  explicit SdnControllerManager(uint64_t device_id) : device_id_(device_id) {}
+
+  grpc::Status HandleArbitrationUpdate(
+      const p4::v1::MasterArbitrationUpdate& update, SdnConnection* controller)
+      ABSL_LOCKS_EXCLUDED(lock_);
```

Comment:
I just realized that we are returning the gRPC status instead of Stratum status here.
In this case, we convert a gRPC status to Stratum status from the caller and convert to another gRPC status and send it back to the controller.
Should we change these functions to use Stratum status?

Reply:
`SDN controller manager` --`grpc::Status`--> `P4Service::AddOrModifyController` --`util::Status`--> `P4Service::StreamChannel` --`grpc::Status`--> gRPC

Reply:
That is somewhat unfortunate, yes. One of the goals with using the external PINS code was getting code that is maintained by them and we can simply update. Modifying it too much or pulling in Stratum deps would make these updates harder.